stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -89,55 +108,93 @@ struct reduce_sum_impl<ReduceFunction, require_var_t<ReturnType>, ReturnType,
         return;
       }
 
-      if (args_adjoints_.size() == 0) {
-        args_adjoints_ = Eigen::VectorXd::Zero(num_vars_shared_terms_);
-      }
-
       // Initialize nested autodiff stack
       const nested_rev_autodiff begin_nest;
 
-      // Create nested autodiff copies of sliced argument that do not point
-      //   back to main autodiff stack
-      std::decay_t<Vec> local_sub_slice;
-      local_sub_slice.reserve(r.size());
-      for (size_t i = r.begin(); i < r.end(); ++i) {
-        local_sub_slice.emplace_back(deep_copy_vars(vmapped_[i]));
+      // Create copies of sliced argument which refer to the main
+      // thread
+      std::decay_t<Vec> local_sub_slice(vmapped_.begin() + r.begin(),
+                                        vmapped_.begin() + r.end());
+
+      // first save current adjoints of sliced argument (which we
+      // restore later) and set these to zero
+      Eigen::VectorXd adjoints_sub_slice(r.size() * num_vars_per_term_);
+      for (std::size_t i = 0; i < r.size() * num_vars_per_term_; ++i) {
+        std::size_t term = r.begin() * num_vars_per_term_ + i;
+        adjoints_sub_slice(i) = varis_[term]->adj_;
+        varis_[term]->adj_ = 0.0;
       }
 
-      // Create nested autodiff copies of all shared arguments that do not point
-      //   back to main autodiff stack
-      auto args_tuple_local_copy = apply(
-          [&](auto&&... args) {
-            return std::tuple<decltype(deep_copy_vars(args))...>(
-                deep_copy_vars(args)...);
-          },
-          args_tuple_);
-
-      // Perform calculation
-      var sub_sum_v = apply(
-          [&](auto&&... args) {
-            return ReduceFunction()(local_sub_slice, r.begin(), r.end() - 1,
-                                    msgs_, args...);
-          },
-          args_tuple_local_copy);
-
-      // Compute Jacobian
-      sub_sum_v.grad();
-
-      // Accumulate value of reduce_sum
-      sum_ += sub_sum_v.val();
-
-      // Accumulate adjoints of sliced_arguments
-      accumulate_adjoints(sliced_partials_ + r.begin() * num_vars_per_term_,
-                          std::move(local_sub_slice));
-
-      // Accumulate adjoints of shared_arguments
-      apply(
-          [&](auto&&... args) {
-            accumulate_adjoints(args_adjoints_.data(),
-                                std::forward<decltype(args)>(args)...);
-          },
-          std::move(args_tuple_local_copy));
+      if (main_thread_ == std::this_thread::get_id()) {
+        // the reducer instance running in the main thread can use the
+        // input operands given as input to reduce_sum which are
+        // stored in the main thread. These need to be zeroed before
+        // use.
+
+        for (std::size_t i = 0; i < num_vars_shared_terms_; ++i) {
+          varis_[vmapped_.size() * num_vars_per_term_ + i]->adj_ = 0.0;
```

Comment:
We would need to save and restore these adjoints like the others

Reply:
The `varis` of the shared arguments are saved and restored in the reduce impl function. The `varis` of the sliced argument is saved, set to 0 and restored in the reduce operation (so that it happens in parallel).

Reply:
Oh I see. Oh right. Yes and this answers the other question too.